cics-bundle-maven-plugin/src/main/java/com/ibm/cics/cbmp/AbstractCICSBundleMojo.java

```diff
@@ -64,36 +70,125 @@
-				return earbundle;
+		} else {
+			//Ignore if it doesn't exist
+		}
+	}
+
+	private void addAutoBundleParts(BundlePublisher bundlePublisher, ArrayList<Artifact> artifacts)
+			throws MojoExecutionException {
+		//For remaining artifacts, find any that are auto-bundle-able
+		getLog().info("Finding project dependencies that can be automatically included in the bundle");
+		for (Artifact artifact : artifacts) {
+			BundlePartBinding bundlePartBinding = getDefaultBundlePartBinding(artifact);
+			if (bundlePartBinding != null) {
+				addBundleResourceForArtifact(bundlePublisher, artifact, bundlePartBinding);
 			}
-			case JAR: {
-				Osgibundle osgibundle = new Osgibundle(getLog());
-				osgibundle.setJvmserver(getDefaultJVMServer());
-				osgibundle.setArtifact(new com.ibm.cics.cbmp.Artifact(a));
-				return osgibundle;
+		}
+	}
+
+	private void addExplicitBundleParts(BundlePublisher bundlePublisher, ArrayList<Artifact> artifacts)
+			throws MojoExecutionException {
+		//From all artifacts, find any that are explicitly bound to specific bundle parts
+		getLog().info("Adding explicitly configured bundle parts");
+		ArrayList<BundlePartBinding> bindings = new ArrayList<>(bundleParts);
+		Iterator<Artifact> artifactsIterator = artifacts.iterator();
+		artifactsLoop: while (artifactsIterator.hasNext()) {
+			Artifact artifact = artifactsIterator.next();
+			Iterator<BundlePartBinding> bindingsIterator = bindings.iterator();
+			while (bindingsIterator.hasNext()) {
+				BundlePartBinding binding = bindingsIterator.next();
+				if (binding.matches(artifact)) {
+					artifactsIterator.remove();
+					bindingsIterator.remove();
+					addBundleResourceForArtifact(bundlePublisher, artifact, binding);
+					continue artifactsLoop;
+				}
 			}
 		}
+		
+		//If there are any bindings that didn't match artifacts, then error!
+		if (!bindings.isEmpty()) {
+			throw new MojoExecutionException("Some bundle part overrides did not correspond to any of the project artifacts: " + bindings.toString());
+		}
+	}
+
+	private void addBundleResourceForArtifact(BundlePublisher bundlePublisher, Artifact artifact,
+			BundlePartBinding binding) throws MojoExecutionException {
+		try {
+			bundlePublisher.addResource(binding.toBundlePart(artifact, this));
+			logBundlePartAdded(artifact, binding);
+		} catch (PublishException e) {
+			throw new MojoExecutionException("Error adding bundle resource for artifact " + artifact.toString() + ": " + e.getMessage());
+		}
+	}
+
+	private void logBundlePartAdded(Artifact artifact, BundlePartBinding bundlePartBinding) {
+		getLog().info("Adding " + bundlePartBinding.getClass().getSimpleName() + " bundle part for " + artifact.getId() + "(" + artifact.getType() + ")");
+	}
+
+	private BundlePartBinding getDefaultBundlePartBinding(Artifact artifact) {
+		switch (artifact.getType()) {
+			case WAR: return new Warbundle();
+			case EAR: return new Earbundle();
+			case JAR: return new Osgibundle();
+			default: return null;
```

Comment:
Should we be throwing some kind of error if the artifact type isn't war, ear or jar?

Reply:
If the artifact's of the wrong type, it just doesn't get added to the bundle unless you've explicitly bound it, so no need to throw an exception